### PR TITLE
✨ Kotlin command builder DSL

### PR DESCRIPTION
**Pull request by jmanpenilla**

### CODE REVIEW

File: cloud-kotlin-extensions/build.gradle [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+compileTestKotlin {
+    kotlinOptions {
+        jvmTarget = "1.8"
+    }
+}
+
+dependencies {
+    api project(':cloud-core')
+    implementation "org.jetbrains.kotlin:kotlin-stdlib-jdk8"
```
Comment (Proximyst) [inline cloud-kotlin-extensions/build.gradle:26]: Don't expect Kotlin versions; let the one who pulls it in do this. `compileOnlyApi` is ideal here.

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt [hunk 2]
```diff
@@ -0,0 +1,363 @@
...
+    /**
+     * Create a new [MutableCommandBuilder]
+     *
+     * @param name name for the root command node
+     * @param description description for the root command node
+     * @param aliases aliases for the root command node
+     * @param commandManager the command manager which will own this command
+     * @since 1.3.0
+     */
+    public constructor(
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:54]: Inline this to the class declaration with inlined variables.

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt [hunk 3]
```diff
@@ -0,0 +1,363 @@
...
+     *
+     * @since 1.3.0
+     */
+    public var senderType: KClass<out C>
+        /**
+         * Write only field.
+         *
+         * @throws UnsupportedOperationException
+         */
+        get() = throw writeOnly
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:205]: Then provide a setter method only.

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt [hunk 4]
```diff
@@ -0,0 +1,363 @@
...
+
+    /**
+     * Specify a required sender type
+     *
+     * @param type sender type
+     * @return this mutable builder
+     * @since 1.3.0
+     */
+    public fun senderType(
+            type: Class<out C>
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:221]: `KClass` too perhaps? Potentially also an `inline fun <reified T : C>`?

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt [hunk 5]
```diff
@@ -0,0 +1,363 @@
...
+     *
+     * @since 1.3.0
+     */
+    public var permission: String
+        /**
+         * Write only field.
+         *
+         * @throws UnsupportedOperationException
+         */
+        get() = throw writeOnly
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:262]: Then provide a setter only.

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt [hunk 6]
```diff
@@ -0,0 +1,363 @@
...
+     *
+     * @since 1.3.0
+     */
+    public var commandPermission: CommandPermission
+        /**
+         * Write only field.
+         *
+         * @throws UnsupportedOperationException
+         */
+        get() = throw writeOnly
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:283]: Then provide a setter only.

File: cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+                            // ...
+                        }
+                    })
+                }
+        )
+
+        manager.buildAndRegister("is") {
+
+            registerCopy {
+                literal("this")
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt:69]: Perhaps we should allow e.g. `registerCopy(literal = "this")`?

File: cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+
+        Assertions.assertEquals(
+                manager.commandHelpHandler.allCommands.map { it.syntaxString }.sorted(),
+                setOf(
+                        "kotlin dsl <moment>",
+                        "kotlin dsl <moment> bruh_moment",
+                        "is",
+                        "is this",
+                        "is this going",
+                        "is this going too_far"
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt:92]: Trailing comments 😠

File: cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+            return SimpleCommandMeta.empty()
+        }
+
+        override fun hasPermission(sender: TestCommandSender, permission: String): Boolean {
+            return !permission.equals("no", ignoreCase = true)
+        }
+    }
+
+    open class TestCommandSender {}
+    class SpecificCommandSender : TestCommandSender() {}
```
Comment (Proximyst) [inline cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt:110]:
```suggestion
    open class TestCommandSender
    class SpecificCommandSender : TestCommandSender()
```

File: cloud-kotlin-extensions/build.gradle [hunk 10]
```diff
@@ -0,0 +1,44 @@
+import org.jetbrains.dokka.gradle.DokkaTask
+
+plugins {
+    id 'org.jetbrains.kotlin.jvm' version '1.4.21'
+    id 'org.jetbrains.dokka' version '1.4.20'
+}
+
+kotlin {
+    explicitApi()
```
Comment (Proximyst) [inline cloud-kotlin-extensions/build.gradle:9]: Didn't know about this! Useful :)

File: cloud-kotlin-extensions/build.gradle [hunk 11]
```diff
@@ -0,0 +1,44 @@
...
+}
+
+javadocJar {
+    from dokkaHtml
+}
+
+tasks.withType(DokkaTask).configureEach {
+    dokkaSourceSets {
+        main {
+            includes.from(layout.projectDirectory.file("src/main/dokka/descriptions.md").toString())
```
Comment (Proximyst) [inline cloud-kotlin-extensions/build.gradle:37]: This should use the resource's source trees I reckon. You have any thoughts?

**Review by Proximyst — CHANGES_REQUESTED** [review]

File: cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt (hunk 4 above)
Comment (jmanpenilla) [reply cloud-kotlin-extensions/src/main/kotlin/cloud/commandframework/kotlin/MutableCommandBuilder.kt:260]: already had these

File: cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt (hunk 7 above)
Comment (jmanpenilla) [reply cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt:69]: added

File: cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt (hunk 8 above)
Comment (jmanpenilla) [reply cloud-kotlin-extensions/src/test/kotlin/cloud/commandframework/kotlin/CommandBuildingDSLTest.kt]: commas

**Pull request merged**